source/Timer.h

```diff
@@ -0,0 +1,126 @@
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef TIMER_H_
+#define TIMER_H_
+
+#include "MissionAction.h"
+#include "System.h"
+
+#include <cstdint>
+#include <map>
+#include <string>
+
+class DataNode;
+class DataWriter;
+class Planet;
+class PlayerInfo;
+class UI;
+
+
+
+// Class representing a timer for triggering mission actions
+// This requires a specification of how long to wait (base + randint(rand)),
+// a system to wait in, and whether to reset the timer upon leaving the system
+// and can optionally require the player to be idle (no acceleration), and/or
+// within or beyond a certain proximity to the system center or a
+// specific (named) system object.
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, const Mission *mission);
+	// Set up the timer from its data file node.
+	void Load(const DataNode &node, const Mission *mission);
+	// Note: the Save() function can assume this is an instantiated Timer, not a template,
+	// so the time to wait will be saved fully calculated, and with any elapsed time subtracted.
+	void Save(DataWriter &out) const;
+
+	// Get whether the timer is optional to complete.
+	bool IsOptional() const;
+	// Get whether the timer is currently active.
+	bool IsComplete() const;
+
+	// Calculate the total time to wait, including any random value.
+	Timer Instantiate(std::map<std::string, std::string> &subs,
+					const System *origin, int jumps, int64_t payload) const;
+	// Progress the timer within the main loop.
+	void Step(PlayerInfo &player, UI *ui, Mission *mission);
+
+
+private:
+	enum class ResetCondition {
+		NONE,
+		PAUSE,
+		LEAVE_ZONE,
+		LEAVE_SYSTEM
+	};
+	void ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, Mission *mission);
+
+
+private:
+	// The default speed for determining whether the player is idle
+	const double DEFAULT_MAX_SPEED = 25.;
+	// The name of the timer.
+	std::string name;
+	// The basic amount of time to wait, with the optional random values.
+	uint32_t timeToWait = 0;
+	uint32_t rand = 0;
+
+	// The system the timer is for.
+	const System *system = nullptr;
+	// The filter for the systems it can be for.
+	LocationFilter systems;
+
+	// If set, the mission the timer is attached to can be completed even without it.
+	bool optional = false;
+	// Whether the timer requires the player to be idle.
+	bool requireIdle = false;
+	// The speed threshold the player's flagship must be under to count as "idle".
+	double idleMaxSpeed = DEFAULT_MAX_SPEED;
```

Comment:
... just give idleMaxSpeed a default value of 25.
```suggestion
	double idleMaxSpeed = 25.;
```